### PR TITLE
harfbuzz: name the DLL on Windows

### PR DESCRIPTION
**Pull request by adrientetar**
r? @ldo

### CODE REVIEW

**Comment by anthrotype** [comment]
The libtool-style dll name you used is the one you get if you compile harfbuzz on Windows with MinGW GCC. If you used Visual Studio the name would be `harfbuzz.dll`.

Also, on OSX the dynamic-load library has `.dylib` extension, not `.so.0`.

Have a look at how other ctypes-based bindings solve the problem of loading the shared library in a way that works on all three operating systems:
https://github.com/rougier/freetype-py/blob/aaf3840/freetype/raw.py#L21-L37

**Comment by adrientetar** [comment]
Humm yes I use the official binaries, would be nice to have MSVC binaries also in 64-bit flavor.

**Comment by ldo** [comment]
Note that version information is important, to handle the case if/when the ABI changes in an incompatible way. That is why I explicitly load `libharfbuzz.so.0`, not `libharfbuzz.so`. All my Python API bindings are careful to do a similar thing, including my [FreeType binding](https://github.com/ldo/python_freetype).

If you don’t get a consistent library name on Windows, then it’s hard to see how to properly support that.

**Comment by anthrotype** [comment]
There is no standard naming convention for DLLs built with MSVC. Anyway, one can always rename the MSVC-built `harfbuzz.dll` to `libharfbuzz-0.dll` (as in this PR), it doesn't really matter.

As far as the macOS dynamic library, you could use the name `libharfbuzz.0.dylib` that contains the ABI version.

**Comment by anthrotype** [comment]
I suggest something like that

```python
if sys.platform == "win32":
    _dll = "libharfbuzz-0.dll"
elif sys.platform == "darwin":
    _dll = "libharfbuzz.0.dylib"
else:
    _dll = "libharfbuzz.so.0"
hb = ct.cdll.LoadLibrary(_dll)
```

**Comment by adrientetar** [comment]
> That is why I explicitly load libharfbuzz.so.0, not libharfbuzz.so.

harfbuzz is backwards compatible since 1.0.0 so, not a big deal afaict.

PR updated.

File: harfbuzz.py [hunk 1]
```diff
@@ -40,7 +41,13 @@
     qahirah = None
 #end try
 
-hb = ct.cdll.LoadLibrary("libharfbuzz.so.0")
+if sys.platform == "win32":
+    _dll = "libharfbuzz-0.dll"
+elif sys.platform == "darwin":
+    _dll = "libharfbuzz.0.dylib"
+else:
+    _dll = "libharfbuzz.so.0"
```
Comment (anthrotype) [inline harfbuzz.py]: you may want to add an `#end if` at the end of the block for consistency with the convention used in the rest of the module. TBH, I believe it's superfluous and haven't seen that anywhere else except in here, but @ldo uses it all the time..

Comment (ldo) [inline harfbuzz.py]: http://ldo17.tumblr.com/post/155087402702/fixing-pythons-greatest-mistake

Comment (anthrotype) [inline harfbuzz.py]: Thanks for the link. Now I understand why you do that. Coming from Python, the indentation rules feel natural and idiomatic to me, while the extra closing comments seem redundant, but I do see your point, and why you'd like to keep it that way in your own code.

**Comment by anthrotype** [comment]
could this be merged now?

**Comment by ldo** [comment]
Given that there is no standard name for the shared library on Windows, why not use the standard Linux name? Or at least, symlink it under that name?

Then I don’t have to change my Python code at all.

**Comment by anthrotype** [comment]
Please reconsider merging this pull request. It does not cost you anything to have an extra if statement in there, and it would make the life of Mac and Windows users much much easier... 🙏

**Comment by ldo** [comment]
If I add code for Windows support, that means I have to field bug reports for Windows, which I have no wish to do.

Both Windows and Mac have their own typography stacks, do they not? HarfPy was created specifically as part of a suite of Python modules to support the Linux typography stack.

**Comment by adrientetar** [comment]
We use harfbuzz on Windows, OSX and GNU/Linux for feature consistency/single code path, the same reasons Mozilla Firefox, Chrome, LibreOffice etc. do.

**Comment by khaledhosny** [comment]
At least a way to set the library name from user code is needed, but merging this change is simpler. I don’t understand the remark about typography stack as well, almost all users of HarfBuzz use it on all platforms to get consistent layout (and because HarfBuzz is so good that it is usually much better than what the platform provides).

**Comment by anthrotype** [comment]
Would you be willing to accept a PR that reads the library path from the environment, e.g. os.environ["HARFBUZZ_LIB"] ?
That would make it easier to embed your module in an application like TruFont.

**Pull request closed without merging**